File: SupermarketMock/SupermarketMock/Models/Coupon.cs
namespace SupermarketMock.Models
{
    public class Coupon
    {
        public int Id { get; set; }
        public string Code { get; set; } = string.Empty;
        public string? Description { get; set; }
        public CouponType Type { get; set; }
        public decimal DiscountValue { get; set; }
        public decimal? MinimumOrderAmount { get; set; }
        public decimal? MaximumDiscountAmount { get; set; }
        public int? UsageLimit { get; set; }
        public int UsedCount { get; set; } = 0;
        public int? UsageLimitPerUser { get; set; }
        public CouponScope Scope { get; set; } = CouponScope.Global;

        public DateTime StartDate { get; set; }
        public DateTime EndDate { get; set; }
        public bool IsActive { get; set; } = true;

        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
        public DateTime? UpdatedAt { get; set; }
        public int? CreatedByUserId { get; set; }

        // Navigation
        public ICollection<CouponUsage> CouponUsages { get; set; } = new List<CouponUsage>();
        public ICollection<CouponProduct> CouponProducts { get; set; } = new List<CouponProduct>();
        public ICollection<CouponCategory> CouponCategories { get; set; } = new List<CouponCategory>();

        // Computed helpers
        public bool IsExpired => DateTime.UtcNow > EndDate;
        public bool IsUsageExhausted => UsageLimit.HasValue && UsedCount >= UsageLimit.Value;
        public bool IsCurrentlyValid => IsActive && !IsExpired && !IsUsageExhausted
                                        && DateTime.UtcNow >= StartDate;
    }

    public enum CouponType
    {
        Percentage,
        FixedAmount,
        FreeShipping
    }

    public enum CouponScope
    {
        Global,
        Product,
        Category
    }
}

File: supermarket-app/src/app/models/coupon.ts
// ===== Enums (matching backend) =====

export enum CouponType {
  Percentage = 'Percentage',
  FixedAmount = 'FixedAmount',
  FreeShipping = 'FreeShipping'
}

export enum CouponScope {
  Global = 'Global',
  Product = 'Product',
  Category = 'Category'
}

// ===== Customer-Facing DTOs =====

export interface CouponListDto {
  id: number;
  code: string;
  description?: string;
  type: CouponType;
  discountValue: number;
  minimumOrderAmount?: number;
  maximumDiscountAmount?: number;
  usageLimit?: number;
  usedCount: number;
  usageLimitPerUser?: number;
  scope: CouponScope;
  startDate: string;
  endDate: string;
  isActive: boolean;
  createdAt: string;
  productIds?: number[];
  categoryIds?: number[];
}

export interface ValidateCouponRequestDto {
  code: string;
  orderSubtotal: number;
  cartProductIds?: number[];
  cartCategoryIds?: number[];
}

export interface CouponValidationResultDto {
  isValid: boolean;
  errorMessage?: string;
  couponId?: number;
  code?: string;
  type?: CouponType;
  discountAmount: number;
  description?: string;
}

export interface ApplyCouponRequestDto {
  code: string;
  orderId: number;
}

export interface CouponUsageDto {
  id: number;
  couponCode: string;
  couponDescription?: string;
  couponType: CouponType;
  discountApplied: number;
  usedAt: string;
  orderId: number;
}

// ===== Admin DTOs =====

export interface CouponStatsDto {
  totalCoupons: number;
  activeCoupons: number;
  expiredCoupons: number;
  totalRedemptions: number;
  totalDiscountGiven: number;
}

export interface CreateCouponDto {
  code: string;
  description?: string;
  type: CouponType;
  discountValue: number;
  minimumOrderAmount?: number;
  maximumDiscountAmount?: number;
  usageLimit?: number;
  usageLimitPerUser?: number;
  scope: CouponScope;
  startDate: string;
  endDate: string;
  isActive: boolean;
  productIds?: number[];
  categoryIds?: number[];
}

export interface UpdateCouponDto extends CreateCouponDto {
  id: number;
  isActive: boolean;
}

// ===== Helper =====

export function getCouponTypeLabel(type: CouponType): string {
  switch (type) {
    case CouponType.Percentage: return 'Percentage';
    case CouponType.FixedAmount: return 'Fixed Amount';
    case CouponType.FreeShipping: return 'Free Shipping';
    default: return 'Unknown';
  }
}

export function getCouponScopeLabel(scope: CouponScope): string {
  switch (scope) {
    case CouponScope.Global: return 'All Products';
    case CouponScope.Product: return 'Specific Products';
    case CouponScope.Category: return 'Specific Categories';
    default: return 'Unknown';
  }
}

export function formatDiscountDisplay(type: CouponType, discountValue: number, maximumDiscountAmount?: number): string {
  switch (type) {
    case CouponType.Percentage:
      let text = `${discountValue}% Off`;
      if (maximumDiscountAmount) {
        text += ` (max HK$${maximumDiscountAmount.toFixed(2)})`;
      }
      return text;
    case CouponType.FixedAmount:
      return `HK$${discountValue.toFixed(2)} Off`;
    case CouponType.FreeShipping:
      return 'Free Shipping';
    default:
      return '';
  }
}